contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+# Resource Classes Proposal
+
+  1. [Abstract](#abstract)
+  2. [Motivation](#motivation)
+  3. [Use Cases](#use-cases)
+  4. [Objectives](#objectives)
+  5. [Non Objectives](#non-objectives)
+  6. [Resource Class](#resource-class)
+  7. [API Changes](#api-changes)
+  8. [Scheduler Changes](#sch-changes)
+  9. [Kubelet Changes](#kubelet-changes)
+ 10. [Opaque Integer Resources](#oir)
+ 11. [Future Scope](#future-scope)
+
+_Authors:_
+
+* @vikaschoudhary16 - Vikas Choudhary &lt;vichoudh@redhat.com&gt;
+* @aveshagarwal - Avesh Agarwal &lt;avagarwa@redhat.com&gt;
+
+## Abstract
+In this document we will describe *resource classes* which is a new model to
+represent compute resources in Kubernetes. This document should be seen as a
+successor to [device plugin proposal](https://github.com/kubernetes/community/pull/695/files)
+and has a dependency on the same.
+
+## Motivation
+Kubernetes system knows only two resource types, 'CPU' and 'Memory'. Any other
+resource can be requested by pod using opaque-integer-resource(OIR) mechanism.
+OIR is a key-value pair with the key being a string and the value being a 'Quantity'
+which can (optionally) be fractional. The current model is great for supporting
+simple compute resources like CPU or Memory, which are available across all
+kubernetes deployments.
+But there is a problem in representing resources like GPUs, ASICs, local storage
+etc in the form of OIRs. Each of such resource type generally has a rich
+metadata like version, capabilities etc. to describe the resource. A particular
+application pod may perform well only with resources which have a certain
+capability, for example GPUs greater than version 'V'. OIR does not allow such,
+metadata based, selection/filtering of resources.
+The current model requires identity mapping between available resources and requested
+resources. Identity mapping means there must be a one-to-one mapping between the
+resource reference used in the spec to request the resource and resource reference
+used to advertise the resource availablity.
```

Comment:
I would highly recommend not mixing quantitative and qualitative concepts. Resources are consumable quantitative things. Capabilities/metadata are not consumable. In other words, you don't "consume an OpenCL capability". Instead, you consume one or more GPUs (or vGPUs) where the thing that is providing the GPU/vGPU has OpenCL/CUDA driver support. The distinction is subtle but important. If you combine the quantitative and qualitative things together, you will end up needing to have many "resource classes" that are only different with regards to the various permutations of capabilities that the provider of the quantitative resources exposes.

For example, you will end up having a bunch of resource classes that read like a vendor product list:

 nvidia.com/gpu/tesla.k80
 nvidia.com/gpu/tesla.k40
 nvidia.com/gpu/tesla.c2075
 nvidia.com/gpu/tesla.c2070
 nvidia.com/gpu/tesla.c2050
 ... lots more for Intel, AMD, etc

Disadvantages of doing things this way include:

* what happens when the inevitable product line rebranding occurs
* you lose the abstraction that should be possible for a cloud to provide with regards to vendor neutrality. Instead of describing capabilities and quantities of like resources, instead the API is exposing vendor-specific stuff
* a lack of flexibility in how users request a quantity of some resource (vGPU or display head, for example) and capabilities (OpenCL support, for example)